launch/launcher.go

```diff
@@ -124,6 +146,86 @@ func (l *Launcher) env(process Process) error {
 	})
 }
 
+func (l *Launcher) execD(process Process) error {
+	execDs, err := l.buildpackFiles(process, ExecDDirName, supportsExecD)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find all exec.d executables in layers dir, '%s'", l.LayersDir)
+	}
+	for _, execD := range execDs {
+		if err := l.ExecD.ExecD(execD, l.Env); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func supportsExecD(bp Buildpack) bool {
+	if bp.API == "" {
+		return false
+	}
+	return api.MustParse(bp.API).Compare(api.MustParse("0.5")) >= 0
+}
+
+func (l *Launcher) buildpackFiles(process Process, dirName string, predicates ...func(Buildpack) bool) ([]string, error) {
+	var files []string
+
+	appendIfFile := func(path string, fi os.FileInfo) {
+		if !fi.IsDir() {
+			files = append(files, path)
+		}
+	}
+
+	appendFilesInDir := func(path string) error {
+		fis, err := ioutil.ReadDir(path)
+		if os.IsNotExist(err) {
+			return nil
+		}
+		if err != nil {
+			return errors.Wrapf(err, "failed to list files in dir '%s'", path)
+		}
+
+		for _, fi := range fis {
+			appendIfFile(filepath.Join(path, fi.Name()), fi)
+		}
+		return nil
+	}
+
+	if err := l.eachBuildpack(func(path string) error {
```

Comment:
This code (and elsewhere where `eachBuildpack` is called) might be easier to understand if we give the function a name, similar to how `buildpackFiles` is called with `supportsExecD`

Reply:
Creating name functions as you suggested turned out to be more difficult than I anticipated because many of these functions manipulated the receiver. I ended up doing a slightly different refactor https://github.com/buildpacks/lifecycle/pull/473/commits/4147c88b565d80820557452e47eb044ae9f6e1cf

Let me know if you think this is more readable. I can back it out if it isn't.

Reply:
Left a couple of suggestions to improve the readability, but overall I think this makes sense!